fedimint-atomic-broadcast/src/broadcast.rs

```diff
@@ -0,0 +1,148 @@
+    db: Database,
+    mut index: u64,
+    mempool_item_receiver: async_channel::Receiver<Vec<u8>>,
+    incoming_message_receiver: async_channel::Receiver<(Message, PeerId)>,
+    outgoing_message_sender: async_channel::Sender<(Message, Recipient)>,
+    ordered_item_sender: mpsc::Sender<OrderedItem>,
+    clean_shutdown_receiver: watch::Receiver<Option<(u64, Duration)>>,
+) -> Shutdown {
+    let (network_data_sender, network_data_receiver) = async_channel::bounded(256);
+    let (signed_block_sender, signed_block_receiver) = async_channel::bounded(16);
+
+    let relay_handle = tokio::spawn(relay_messages(
+        db.clone(),
+        incoming_message_receiver.clone(),
+        outgoing_message_sender.clone(),
+        network_data_sender.clone(),
+        signed_block_sender.clone(),
+    ));
+
+    loop {
+        if let Some(signed_block) = db::load_block(&db, index).await {
+            tracing::info!("Loaded block with index {}", index);
+            if process_signed_block(signed_block, ordered_item_sender.clone())
+                .await
+                .is_err()
+            {
+                relay_handle.abort();
+                relay_handle.await.ok();
+
+                return Shutdown::MidSession(index);
+            }
+        } else {
+            tracing::info!("Run session with index {}", index);
+            let (backup_loader, backup_saver) = db::open_session(db.clone(), index).await;
+
+            let session_result = session::run(
+                index,
+                keychain.clone(),
+                backup_loader,
+                backup_saver,
+                mempool_item_receiver.clone(),
+                network_data_receiver.clone(),
+                outgoing_message_sender.clone(),
+                ordered_item_sender.clone(),
+                signed_block_receiver.clone(),
+            );
+
+            match session_result.await {
+                Ok(signed_block) => {
+                    tracing::info!("Completed session with index {}", index);
+                    db::complete_session(&db, index, signed_block.clone()).await;
+
+                    outgoing_message_sender
+                        .send((Message::Block(signed_block), Recipient::Everyone))
+                        .await
+                        .ok();
+                }
+                Err(..) => {
+                    relay_handle.abort();
+                    relay_handle.await.ok();
+
+                    return Shutdown::MidSession(index);
+                }
+            };
+        }
+
+        let clean_shutdown = clean_shutdown_receiver.borrow().to_owned();
+        if let Some((shutdown_index, shutdown_delay)) = clean_shutdown {
+            if index == shutdown_index {
+                tracing::info!("Initiate clean shutdown after index {}", index);
+
+                // we delay the shutdown to allow lagging nodes to complete the session as well
+                tokio::time::sleep(shutdown_delay).await;
```

Comment:
Could this result in inconsistent saved state?

Reply:
Do you mean complete_session? In that case I don't think so as we remove the aleph backup buffer and write the signed block to the db in a single transaction. If we have a signed block in the database we do not open the session again if that is what you are concerned by...?